File: FindMinimuminRotatedSortedArray.swift

//After class
// Time Complexity : (O)logn
// Space Complexity : (O)1
// Did this code successfully run on Leetcode : Yes

//Use binary search to find pivot element
// use two pointer technique. keep track of low and high pointer
// if mid element is less than mid-1 and mid+1 then return mid element
// otherwise move the high pointer to mid -1 when mid value is less than high pointer value
// Or move the low pointer to mid + 1 when mid value is higher than high pointer value

class Solution {
    func findMin(_ nums: [Int]) -> Int {
       //edge
       guard nums.count > 0 else { return -1 }
       
       var low = 0
       var high = nums.count - 1
       while low <= high {
           if nums[low] <= nums[high] {
               return nums[low]
           }
           let mid = low + (high - low)/2
           if (mid == 0 || nums[mid] < nums[mid - 1]) && 
              (mid == nums.count - 1 || nums[mid] < nums[mid + 1]) {
                  return nums[mid]
           } else if nums[mid] < nums[high] {
                high = mid - 1
           } else {
                low = mid + 1
           }
       } 
       return -1 
    }
}

File: findMinimumInRotatedSortedArray.cpp
//time complexity: O(n)
//space complexity:O(1)
//approach: binary search, found the target first, which is the minimum element in array, then used binary search to find the index of that array
//solved on leetcode? yes
//problem faced? yes, it is a brute force approach. Although I'm using binary search but still the complexity will be O(n)
//because I scanned the array linearly to find the min element which is O(n)
class Solution {
public:
    int findMin(vector<int>& nums) {
        int target=INT_MAX;
        for(int i=0; i<nums.size(); i++)
        {
            if(nums[i]<target)
            {
                target=nums[i];
            }
        }
        int l=0;
        int r=nums.size()-1;
        while(l<=r)
        {
            int mid=l+(r-l)/2;
            if(nums[mid]==target)
                return nums[mid];
            if(nums[l]<=nums[mid])
            {
                if(nums[l]<=target && target<nums[mid])
                    r=mid-1;
                else
                    l=mid+1;
            }
            else
            {
                if(nums[mid]<target && target<=nums[r])
                    l=mid+1;
                else
                    r=mid-1;
            }
        }
        return -1;
    }
};



//Approach -2

class Solution {
public:
    int findMin(vector<int>& nums) {
        int l=0;
        int r=nums.size()-1;
        if(nums[r]>nums[l] || nums.size()==1)
            return nums[l];
        while(l<=r)
        {
            int mid=l+(r-l)/2;
            if(nums[mid]>nums[mid+1])
            {
                return nums[mid+1];
            }
            else if(nums[mid-1]>nums[mid])
            {
                return nums[mid];
            }
            if(nums[mid]>nums[0])
            {
                l=mid+1;
            }
            else
            {
                r=mid-1;
            }
        }
        return -1;
    }
};